idaes/apps/caprese/data_manager.py

```diff
@@ -0,0 +1,361 @@
+                                            self.plantblock.differential_vars + \
+                                                self.plantblock.input_vars
+
+        self.plant_df = empty_dataframe_from_variables(self.plant_vars_of_interest)
+        
+    def get_plant_dataframe(self):
+        return self.plant_df
+        
+    def save_initial_plant_data(self):
+        plant_states_to_save = self.plant_user_interested_states + self.plantblock.differential_vars
+        self.plant_df = add_variable_values_to_dataframe(self.plant_df,
+                                                         plant_states_to_save, #no inputs
+                                                         iteration = 0,
+                                                         time_subset = [self.plantblock.time.first()])
+        
+
+    def save_plant_data(self, iteration):
+        #skip time.first()
+        time_subset = self.plantblock.time.ordered_data()[1:]
+        self.plant_df = add_variable_values_to_dataframe(self.plant_df,
+                                                         self.plant_vars_of_interest,
+                                                         iteration,
+                                                         time_subset = time_subset,)
+        
+class ControllerDataManager(PlantDataManager, NMPC_PlotLibrary):
+    def __init__(self, 
+                 plantblock, 
+                 controllerblock, 
+                 user_interested_states = [],
+                 user_interested_inputs = []):
+        
+        if not hasattr(self, "plant_df"):
+            super(ControllerDataManager, self).__init__(plantblock, 
+                                                        user_interested_states, 
+                                                        user_interested_inputs,)
+        
+        self.controllerblock = controllerblock
+        # Convert vars in plant to vars in controller
+        if user_interested_states != []:
+            cuid_states = [ComponentUID(var.referent) for var in user_interested_states]
+            self.controller_user_interested_states = [cuid.find_component_on(controllerblock)
+                                                      for cuid in cuid_states]
+        else:
+            self.controller_user_interested_states = []
+            
+        if user_interested_inputs != []:
+            cuid_inputs = [ComponentUID(var.referent) for var in user_interested_inputs]
+            self.controller_user_interested_inputs = [cuid.find_component_on(controllerblock)
+                                                      for cuid in cuid_inputs]
+        else:
+            self.controller_user_interested_inputs = []
+        
+        
+        self.controller_vars_of_interest = self.controller_user_interested_inputs + \
+                                                controllerblock.input_vars   
+        self.controller_df = empty_dataframe_from_variables(self.controller_vars_of_interest)
```

Comment:
Ah, I didn't think that far. It's just because if you print `ComponentUID(var.referent)` for the variable in the list `input_vars`, it might get "double" asterisks for the input that only have time index. In that case, user-given inputs cannot be found because there is only one asterisk in their CUID. I think it's bug for CUID, so we have this kind of redundant part.

Reply:
Okay, I think I see. I'll look at this a little closer, but I think this is a bug in the `pyomo.dae.flatten` that I should fix.

Reply:
Ok, thanks! Let me know when you fix it. Then I can modify this function.